presto-main/src/main/java/com/facebook/presto/execution/QueueableExecution.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.concurrent.SetThreadName;
+
+import java.util.List;
+import java.util.concurrent.Executor;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class QueueableExecution
+{
+    private final QueryExecution queryExecution;
+    private final List<QueryQueue> nextQueues;
+    private final ListenableFuture<?> listenableFuture;
+    private final Executor executor;
+    private final SqlQueryManagerStats stats;
+
+    public QueueableExecution(QueryExecution queryExecution, List<QueryQueue> nextQueues, Executor executor, SqlQueryManagerStats stats)
+    {
+        this.queryExecution = checkNotNull(queryExecution, "queryExecution is null");
+        this.nextQueues = ImmutableList.copyOf(checkNotNull(nextQueues, "nextQueues is null"));
+        this.executor = checkNotNull(executor, "executor is null");
+        this.stats = checkNotNull(stats, "stats is null");
+        SettableFuture<?> settableFuture = SettableFuture.create();
+        SqlQueryManager.addCompletionCallback(queryExecution, () -> settableFuture.set(null));
```

Comment:
I don't think it is, actually. Lambdas behave differently from anonymous inner classes, and only capture `this` if it's referenced by the body.

I couldn't find the relevant section in the JLS, but see: http://cr.openjdk.java.net/~briangoetz/lambda/lambda-state-4.html

"while inner class instances always hold a strong reference to their enclosing instance, lambdas that do not capture members from the enclosing instance do not hold a reference to it."